File: zeromq.py
# CHANGES - functions naming convention changed to remove 'zeromq'.  As we are now using imports, functions will be
# be called using [filename].[function] style so instead of 'zeromq_start()' we will have 'zeromq.start()'.  Variable
# and function naming also changed to conform to PEP8 standards (no camelcase, underscores separating words when
# appropriate)

import json
import zmq
import chess
from chessai import ChessAI

zeromq_bool_running = False


def start(main_int_zeromq, main_str_name):
    global zeromq_bool_running

    zeromq_bool_running = True

    context_handle = zmq.Context()
    socket_handle = context_handle.socket(zmq.PAIR)

    socket_handle.bind('tcp://*:' + str(main_int_zeromq))

    # CHANGES - json_out has a static number of fields that it modifies as needed to pass information back to the
    # framework.  For this reason, it never needs to be deleted and instead can be set up with all required keys before
    # entering the main loop.  This will save us some computation time as we never have to recreate the json_out
    # dictionary and dictionary assignment to already existing keys is very very fast.
    # After the below call to dict.fromkeys() all the values will be None.
    json_out = dict.fromkeys(['strOut', 'strReturn', 'boolReturn', 'intReturn', 'intOut'])

    ai = ChessAI()

    while zeromq_bool_running:
        # CHANGES - json_in variable does not need to be set to None at the end of the file as this call will overwrite
        # the previous object and cause it to be garbage collected just the same.
        json_in = json.loads(socket_handle.recv().decode())

        if json_in['strFunction'] == 'ping':
            json_out['strOut'] = main_str_name

        elif json_in['strFunction'] == 'chess_reset':
            # chess.reset()
            ai.reset()

        elif json_in['strFunction'] == 'chess_boardGet':
            # json_out['strOut'] = chess.board_get()
            json_out['strOut'] = ai.board_get()

        elif json_in['strFunction'] == 'chess_boardSet':
            # chess.board_set(json_in['strIn'])
            ai.board_set(json_in['strIn'])

        elif json_in['strFunction'] == 'chess_winner':
            # json_out['strReturn'] = chess.winner()
            json_out['strReturn'] = ai.winner()

        elif json_in['strFunction'] == 'chess_isValid':
            # json_out['boolReturn'] = chess.is_valid(json_in['intX'], json_in['intY'])
            json_out['boolReturn'] = ChessAI.is_valid(json_in['intY'], json_in['intX'])

        elif json_in['strFunction'] == 'chess_isEnemy':
            # json_out['boolReturn'] = chess.is_enemy(json_in['strPiece'])
            json_out['boolReturn'] = ai.is_enemy(json_in['strPiece'])

        elif json_in['strFunction'] == 'chess_isOwn':
            # json_out['boolReturn'] = chess.is_own(json_in['strPiece'])
            json_out['boolReturn'] = ai.is_own(json_in['strPiece'])

        elif json_in['strFunction'] == 'chess_isNothing':
            # json_out['boolReturn'] = chess.is_nothing(json_in['strPiece'])
            json_out['boolReturn'] = ChessAI.is_nothing(json_in['strPiece'])

        elif json_in['strFunction'] == 'chess_eval':
            # json_out['intReturn'] = chess.eval()
            json_out['intReturn'] = ai.eval()

        elif json_in['strFunction'] == 'chess_moves':
            # str_out = chess.moves()
            # json_out['intOut'] = len(str_out)
            # CHANGES - this is a more Pythonic way of calling join
            # json_out['strOut'] = ''.join(str_out)
            # str_out = ai.framework_moves()
            str_out = ai.framework_moves()
            json_out['intOut'] = len(str_out)
            json_out['strOut'] = ''.join(str_out)

        elif json_in['strFunction'] == 'chess_movesShuffled':
            str_out = chess.moves_shuffled()

            json_out['intOut'] = len(str_out)
            # CHANGES - this is a more Pythonic way of calling join
            json_out['strOut'] = ''.join(str_out)

        elif json_in['strFunction'] == 'chess_movesEvaluated':
            str_out = chess.moves_evaluated()

            json_out['intOut'] = len(str_out)
            # CHANGES - this is a more Pythonic way of calling join
            json_out['strOut'] = ''.join(str_out)

        elif json_in['strFunction'] == 'chess_move':
            # chess.move(json_in['strIn'])
            ai.framework_move(json_in['strIn'])

        elif json_in['strFunction'] == 'chess_moveRandom':
            json_out['strOut'] = chess.move_random()

        elif json_in['strFunction'] == 'chess_moveGreedy':
            json_out['strOut'] = chess.move_greedy()

        elif json_in['strFunction'] == 'chess_moveNegamax':
            json_out['strOut'] = chess.move_negamax(json_in['intDepth'], json_in['intDuration'])

        elif json_in['strFunction'] == 'chess_moveAlphabeta':
            json_out['strOut'] = chess.move_alphabeta(json_in['intDepth'], json_in['intDuration'])

        elif json_in['strFunction'] == 'chess_undo':
            chess.undo()

        socket_handle.send(json.dumps(json_out).encode())

    socket_handle.close()
    context_handle.close()


def stop():
    global zeromq_bool_running

    zeromq_bool_running = False

File: zeromq.c
#include "zeromq.h"

#include <unistd.h>

#include "main.h"
#include "finterface.h"

bool zeromq_boolRunning = false;

void zeromq_start() {
    {
        zeromq_boolRunning = true;
    }
    
    {
        void* contextHandle = zmq_ctx_new();
        void* socketHandle = zmq_socket(contextHandle, ZMQ_PAIR);
        
        {
            char charZeromq[1024] = { };
            
            sprintf(charZeromq, "tcp://*:%d", main_intZeromq);
            
            if (zmq_bind(socketHandle, charZeromq) == -1) {
                printf("zeromq: %s\n", zmq_strerror(zmq_errno()));
            }
        }
        
        {
            ChessAI ai;
            ChessAI_init(&ai);
            do {
                cJSON* cjsonIn = NULL;
                cJSON* cjsonOut = cJSON_CreateObject();
                
                {
                    char charJson[1024] = { };
                    
                    if (zmq_recv(socketHandle, charJson, sizeof(charJson), 0) == -1) {
                        printf("zeromq: %s\n", zmq_strerror(zmq_errno()));
                    }
                    
                    cjsonIn = cJSON_Parse(charJson);
                }
                
                {
                    if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "ping") == 0) {
                        cJSON_AddStringToObject(cjsonOut, "strOut", main_charName);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_reset") == 0) {
                        // chess_reset();
                        frameworkInterface_reset(&ai);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_boardGet") == 0) {
                        char charOut[1024] = { };
                        
                        // chess_boardGet(charOut);
                        frameworkInterface_getBoard(&ai, charOut);
                        
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_boardSet") == 0) {
                        char charIn[1024] = { };
                        
                        strcpy(charIn, cJSON_GetObjectItem(cjsonIn, "strIn")->valuestring);
                        
                        // chess_boardSet(charIn);
                        frameworkInterface_setBoard(&ai, charIn);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_winner") == 0) {
                        char charReturn[16] = { };
                        
                        // charReturn[0] = chess_winner();
                        charReturn[0] = frameworkInterface_winner(&ai);
                        charReturn[1] = '\0';
                        
                        cJSON_AddStringToObject(cjsonOut, "strReturn", charReturn);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_isValid") == 0) {
                        // cJSON_AddBoolToObject(cjsonOut, "boolReturn", chess_isValid(cJSON_GetObjectItem(cjsonIn, "intY")->valueint, cJSON_GetObjectItem(cjsonIn, "intX")->valueint));
                        cJSON_AddBoolToObject(cjsonOut, "boolReturn", frameworkInterface_isValid(cJSON_GetObjectItem(cjsonIn, "intY")->valueint, cJSON_GetObjectItem(cjsonIn, "intX")->valueint));
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_isEnemy") == 0) {
                        char charIn[16] = { };
                        
                        strcpy(charIn, cJSON_GetObjectItem(cjsonIn, "strPiece")->valuestring);
                        
                        // cJSON_AddBoolToObject(cjsonOut, "boolReturn", chess_isEnemy(charIn[0]));
                        cJSON_AddBoolToObject(cjsonOut, "boolReturn", frameworkInterface_isEnemy(&ai, charIn[0]));
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_isOwn") == 0) {
                        char charIn[16] = { };
                        
                        strcpy(charIn, cJSON_GetObjectItem(cjsonIn, "strPiece")->valuestring);
                        
                        // cJSON_AddBoolToObject(cjsonOut, "boolReturn", chess_isOwn(charIn[0]));
                        cJSON_AddBoolToObject(cjsonOut, "boolReturn", frameworkInterface_isOwn(&ai, charIn[0]));
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_isNothing") == 0) {
                        char charIn[16] = { };
                        
                        strcpy(charIn, cJSON_GetObjectItem(cjsonIn, "strPiece")->valuestring);
                        
                        // cJSON_AddBoolToObject(cjsonOut, "boolReturn", chess_isNothing(charIn[0]));
                        cJSON_AddBoolToObject(cjsonOut, "boolReturn", frameworkInterface_isNothing(&ai, charIn[0]));
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_eval") == 0) {
                        // cJSON_AddNumberToObject(cjsonOut, "intReturn", chess_eval());
                        cJSON_AddNumberToObject(cjsonOut, "intReturn", frameworkInterface_eval(&ai));
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_moves") == 0) {
                        int intOut = 0;
                        char charOut[1024] = { };
                        
                        // intOut = chess_moves(charOut);
                        intOut = frameworkInterface_moves(&ai, charOut);
                        
                        cJSON_AddNumberToObject(cjsonOut, "intOut", intOut);
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_movesShuffled") == 0) {
                        int intOut = 0;
                        char charOut[1024] = { };
                        
                        // intOut = chess_movesShuffled(charOut);
                        intOut = frameworkInterface_movesShuffled(&ai, charOut);
                        
                        cJSON_AddNumberToObject(cjsonOut, "intOut", intOut);
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_movesEvaluated") == 0) {
                        int intOut = 0;
                        char charOut[1024] = { };
                        
                        // intOut = chess_movesEvaluated(charOut);
                        intOut = frameworkInterface_movesEvaluated(&ai, charOut);
                        
                        cJSON_AddNumberToObject(cjsonOut, "intOut", intOut);
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_move") == 0) {
                        char charIn[16] = { };
                        
                        strcpy(charIn, cJSON_GetObjectItem(cjsonIn, "strIn")->valuestring);
                        
                        // chess_move(charIn);
                        frameworkInterface_move(&ai, charIn);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_moveRandom") == 0) {
                        char charOut[16] = { };
                        
                        // chess_moveRandom(charOut);
                        frameworkInterface_moveRandom(&ai, charOut);
                        
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_moveGreedy") == 0) {
                        char charOut[16] = { };
                        
                        // chess_moveGreedy(charOut);
                        frameworkInterface_moveGreedy(&ai, charOut);
                        
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_moveNegamax") == 0) {
                        char charOut[16] = { };
                        
                        // chess_moveNegamax(charOut, cJSON_GetObjectItem(cjsonIn, "intDepth")->valueint, cJSON_GetObjectItem(cjsonIn, "intDuration")->valueint);
                        frameworkInterface_moveNegamax(&ai, charOut, cJSON_GetObjectItem(cjsonIn, "intDepth")->valueint, cJSON_GetObjectItem(cjsonIn, "intDuration")->valueint);
                        
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_moveAlphabeta") == 0) {
                        char charOut[16] = { };
                        
                        // chess_moveAlphabeta(charOut, cJSON_GetObjectItem(cjsonIn, "intDepth")->valueint, cJSON_GetObjectItem(cjsonIn, "intDuration")->valueint);
                        frameworkInterface_moveAlphabeta(&ai, charOut, cJSON_GetObjectItem(cjsonIn, "intDepth")->valueint, cJSON_GetObjectItem(cjsonIn, "intDuration")->valueint);
                        
                        cJSON_AddStringToObject(cjsonOut, "strOut", charOut);
                        
                    } else if (strcmp(cJSON_GetObjectItem(cjsonIn, "strFunction")->valuestring, "chess_undo") == 0) {
                        // chess_undo();
                        frameworkInterface_undo(&ai);
                        
                    }
                }
                
                {
                    char* charJson = cJSON_PrintUnformatted(cjsonOut);
                    
                    if (zmq_send(socketHandle, charJson, strlen(charJson), 0) == -1) {
                        printf("zeromq: %s\n", zmq_strerror(zmq_errno()));
                    }
                    
                    free(charJson);
                }
                
                cJSON_Delete(cjsonIn);
                cJSON_Delete(cjsonOut);
            } while (zeromq_boolRunning == true);

            // ChessAI_destroy(&ai);
        }
        
        zmq_close(socketHandle);
        zmq_ctx_destroy(contextHandle);
    }
}

void zeromq_stop() {
    {
        zeromq_boolRunning = false;
    }
}
